odl/solvers/advanced/chambolle_pock.py

```diff
@@ -0,0 +1,278 @@
+       operator of ``G``. Domain and range of the returned proximal operator
+       is the domain of ``K``.
+    sigma : positive `float`
+        Convergence proofed for ``||K||_2^2 * sigma * tau < 1``
+    tau : positive `float`
+        Convergence proofed for ``||K||_2^2 * sigma * tau < 1``
+    theta : `float` in [0, 1], optional
+        Relaxation paramter
+    niter : `int`, optional
+        Number of iterations
+    partial : `Partial`, optional
+        If not `None` the object passed executes code per iteration,
+        e.g. plotting each iterate
+
+    Returns
+    -------
+    x : `DiscreteLpVector`
+
+    References
+    ----------
+    Original paper by `Chambolle and Pock`_. Also see `diagonal
+    preconditioning`_ techniques. This implementation is based on the
+    article on convex optimization problem prototyping for image
+    reconstruction in computed tomography by `Sidky et al.`_.
+
+    .. _Chambolle and Pock: http://dx.doi.org/10.1007/s10851-010-0251-1
+    .. _diagonal preconditioning: http://dx.doi.org/10.1109/ICCV.2011.6126441
+    .. _Sidky et al.: http://stacks.iop.org/0031-9155/57/i=10/a=3065
+    """
+    # Initialize the primal and dual variable
+    x = K.domain.zero()
+    y = K.range.zero()
+    # Relaxation variable
+    xbar = x.copy()
+
+    # Initialize the proximal operator of the convex conjugate of functional F
+    f_cc_prox_sigma = f_cc_prox(sigma)
+    # Initialize the proximal operator of functional G
+    g_prox_tau = g_prox(tau)
+    # Adjoint of the product space operator
+    Kadjoint = K.adjoint
+
+    for nn in range(niter):
+        # Copy required for relaxation
+        xold = x.copy()
+        # Gradient descent in the dual variable y
+        f_cc_prox_sigma(y + sigma * K(xbar), out=y)
+        # Gradient descent in the primal variable x
+        g_prox_tau(x - tau * Kadjoint(y), out=x)
+        # Overrelaxation in the primal variable x
+        xbar = x + theta * (x - xold)
+
+        if partial is not None:
+            partial.send(x)
+
+    return x
+
+
+def f_cc_prox_l2_tv(space, g, lam):
+    """Function for the proximal operator with l2-data plus TV-regularization.
+
+    Factory function which provides a function to initialize the proximal
+    operator of the convex conjugate of the functional ``F`` given by the
+    L2-data term and the isotropic total variation semi-norm as in the primal
+    minimization problem:
+
+    F(y,z)= 1/2 * ||y - g||_2^2 + lambda * ||(|grad u|)||_1
+
+    with y = Au and z = grad u. The operators ``A`` and ``grad`` are
+    combined to a matrix operator ``K`` as:
+
+        K = (A, grad)^T
+
+    The proximal operator is mapping from a vector space X to its dual space
+    X^*. We assume X to be Hilbert spaces which is self-dual. Here, the domain
+    and range of the proximal operator are given by the range of ``K``.
+
+    Parameters
+    ----------
+    space : `ProductSpace`
```

Comment:
Why does the user need to provide this? Shouldnt the domain of the `A` operator be sufficient?